### PR TITLE
Tpetra: Call KokkosSparse::spgemm with integer row pointers

### PR DESCRIPTION
**Pull request by cwpearson**
Adds a pathway to call KokkosSparse::spgemm in MatrixMatrix with integer-typed row pointer arrays, for better TPL support.

> [!NOTE]
> This can mostly be removed once Tpetra matrices default to int/int.

Builds on the approach in [\#12852](https://github.com/trilinos/Trilinos/pull/12852)

- [x] cuSparse
  - [x] patch-match https://github.com/kokkos/kokkos-kernels/pull/2213 (@ndellingwood)
- [x] rocSparse
- [x] SYCL
- forward-declare several parts of the MatrixMatrix implementation, since `CrsMatrix` has to friend it (and move default parameters to forward declaration)
- adds a private `getApplyHelper()` method to `CrsMatrix` for the MatrixMatrix implementation to get access to the cached integer row pointer
* Introduces `Tpetra::is_crs_matrix_v<Matrix>`, `true` iff Matrix is a CrsMatrix specialization.
* Introduces `Tpetra::Details::IntRowPtrHelper`, a struct to help lazily generate and cache integer-typed row pointers for `CrsMatrix` and `BlockCrsMatrix` (`ApplyMatrixHelper` now uses this)

### CODE REVIEW

File: packages/tpetra/core/test/MatrixMatrix/MatrixMatrix_UnitTests.cpp [hunk 1]
```diff
@@ -2849,7 +2849,7 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL(Tpetra_MatMatAdd, locally_unsorted, SC, LO, GO
 // FIXME_SYCL requires querying free device memory in KokkosKernels, see
 // https://github.com/kokkos/kokkos-kernels/issues/1062.
 // The SYCL specifications don't allow asking for that.
-#ifdef HAVE_TPETRA_SYCL
+#if defined(HAVE_TPETRA_SYCL) || 1
```
[the remark this answers is not in the document]
Comment (cwpearson) [reply packages/tpetra/core/test/MatrixMatrix/MatrixMatrix_UnitTests.cpp:2852]: Separate PR for this: https://github.com/trilinos/Trilinos/pull/13198

File: packages/tpetra/core/test/MatrixMatrix/FECrs_MatrixMatrix_UnitTests.cpp [hunk 2]
```diff
@@ -394,7 +394,7 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL (Tpetra_MatMat, FECrsMatrix, SC, LO, GO, NT)
   TPETRA_ETI_MANGLING_TYPEDEFS()
 
 // FIXME_SYCL
-#ifndef KOKKOS_ENABLE_SYCL
+#if !defined(KOKKOS_ENABLE_SYCL) || 1
```
[the remark this answers is not in the document]
Comment (cwpearson) [reply packages/tpetra/core/test/MatrixMatrix/FECrs_MatrixMatrix_UnitTests.cpp:397]: Separate PR for this: https://github.com/trilinos/Trilinos/pull/13198

File: packages/muelu/test/unit_tests/Regression.cpp [hunk 3]
```diff
@@ -103,17 +105,26 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL(Regression, H2D, Scalar, LocalOrdinal, GlobalO
   // confirm that we did get a hierarchy with two levels -- a sanity check for this test
   TEST_EQUALITY(2, H->GetGlobalNumLevels());
 
+  // When Kokkos Kernels uses TPLs, some Kokkos::deep_copy in the Kokkos Kernels native implementations are not called.
+#if  defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) \
+  || defined(KOKKOSKERNELS_ENABLE_TPL_ROCSPARSE) \
+  || defined(KOKKOSKERNELS_ENABLE_TPL_MKL)
+  constexpr int kkNativeDeepCopies = 0;
```
Comment (cgcgcg) [inline packages/muelu/test/unit_tests/Regression.cpp:112]: This probably needs to be change into something like
```
#if  defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE)
if (Tpetra node is Cuda)
```
and so on. In general this rarely matters, but some of the nightlies might be built with Cuda+Serial enabled.

**Comment by jhux2** [comment]
The tests `TpetraCore_MatrixMatrix_UnitTests_MPI_4` fail on Sunspot (testbed for Aurora).  I'm attaching the verbose output.  (I added extraneous printf debugging a while ago, sorry about the extra lines).

[trilinos_unit_tests.o9016064.txt](https://github.com/user-attachments/files/16198784/trilinos_unit_tests.o9016064.txt)

**Comment by jhux2** [comment]
@brian-kelley said that the MKL that ships with Intel OneAPI 2024.1 may fix this, but that some cmake changes are necessary.

**Comment by brian-kelley** [comment]
To add to what @jhux2 said, #12941 was my first attempt to get this working and is still pretty close to what we need for this new OneAPI version (the relevant changes are in cmake/TPLs). I'll finish this up when I get a chance

**Comment by cwpearson** [comment]
@csiefer2 if I'm understanding the dashboard correctly, only a non-blocking test failed. Merge?

**Comment by jhux2** [comment]
> @csiefer2 if I'm understanding the dashboard correctly, only a non-blocking test failed. Merge?

@cwpearson Nope, cannot, merging is blocked.  I think re-testing is the only option.

**Comment by jhux2** [comment]
Issue might have been the same as https://github.com/trilinos/Trilinos/pull/13263#issuecomment-2251053714.

**Comment by cwpearson** [comment]
Oh yeah, I see now that even the non-blocking failure is actually an old one. I will try to retest I guess.

**Pull request merged**